### PR TITLE
Add Reflect.enumerateClasses to enumerate all classes selected with reflectEnumerateClass().

### PR DESCRIPTION
**Pull request by AugustNagro**

### CODE REVIEW

**Comment by sjrd** [comment]
What is the difference between `getClassAndDescendents()` and the existing `getClassForName()`?

File: src/main/scala/be/doeraene/sjsreflect/Reflect.scala [hunk 1]
```diff
@@ -45,6 +42,13 @@ object Reflect {
   def getClassForName(fqcn: String): Option[Class[_]] =
     classesByName.get(fqcn)
 
+  /**
+    * Get this class and its descendants. Requires the `selectDescendantsClasses` selector
+    * @param fqcn the class's fully qualified name
+    * @return a js.Dictionary of class name -> Class[_]
+    */
+  def getClassAndDescendants(fqcn: String): Dictionary[Class[_]] = classesByName
```
Comment (sjrd) [inline src/main/scala/be/doeraene/sjsreflect/Reflect.scala]: The fact that this method takes a `fqcn` as argument but does not use it looks suspicious ;)

Comment (AugustNagro) [reply src/main/scala/be/doeraene/sjsreflect/Reflect.scala]: My bad. Would you like me to remove the argument and rename to "getAllClassesAndDescendents"?

**Comment by AugustNagro** [comment]
literally the only difference is that it returns the full dictionary.

**Comment by sjrd** [comment]
So the use case is that you want to iterate over all classes on the classpath?

**Comment by AugustNagro** [comment]
Yup. For each class extending a trait (that is selected by the reflection plugin), I'm injecting some js in the launcher's constructor.

**Comment by sjrd** [comment]
Hum. I think this should be a separate capability than `getClassForName`. It is both more powerful (it provides enumeration) and less powerful (it does not really need the _names_, only the `Class`es).

So I believe there should be a separate `Operation` capability, which would be `ReflectEnumerateClass`. That would put the classes it applies to in a separate _array_ (not a dictionary) of `java.lang.Class`es. That array would be accessed through `Reflect.enumerateClasses()`.

**Comment by AugustNagro** [comment]
It looks like 2/3 of the tests failed when downloading dependencies.

File: .gitignore [hunk 2]
```diff
@@ -4,3 +4,4 @@ bin/
 .classpath
 .project
 .settings/
+.idea/
```
Comment (sjrd) [inline .gitignore]: This should be part of your global git-ignore file, not in local projects. At the very least, it should not be part of this PR, so please revert this change.

File: README.md [hunk 3]
```diff
@@ -49,14 +49,17 @@ at the end of its full name, e.g., `"foo.Bar$"`.
 Available operations are:
 
 * `reflectClassByName()`: enables `Reflect.getClassForName`
-* `reflectDeclaredConstructors()`: enables `Reflect.getDeclaredConstructors`
+* `reflectEnumerateClass()` enables `Reflect.enumerateClasses`
+* `reflectDeclaredConstructors()`: enables `Reflect.getDeclaredConstructors` and `Reflect.getAllClasses`
```
Comment (sjrd) [inline README.md]: "and `Reflect.getAllClasses`" is not relevant anymore.

File: README.md [hunk 4]
```diff
@@ -49,14 +49,17 @@ at the end of its full name, e.g., `"foo.Bar$"`.
...
 * `reflectModuleAccessor()`: enables `Reflect.loadModule`
 
 ## API
 
 * `be.doeraene.sjsreflect.Reflect`
   * `Reflect.getClassForName(fqcn: String): Class[_]`:
     find a class by its name
+* `Reflect.enumerateClasses: js.Array[Class[_]]`:
```
Comment (sjrd) [inline README.md]: 2 spaces are missing at the beginning of this line.

File: README.md [hunk 5]
```diff
@@ -49,14 +49,17 @@ at the end of its full name, e.g., `"foo.Bar$"`.
...
+    enumerate all classes specified by the `reflectEnumerateClass()` operator (and descendants if the  `selectDescendentClasses` selector is used)
```
Comment (sjrd) [inline README.md]: operator -> operation (as used elsewhere in the readme)

The part "(and descendants if the `selectDescendentClasses` selector is used)" is redundant. It is a consequence of how selectors work. There is no reason to add this here but not in other operations; and we definitely don't want to repeat it on every operation. So it should be removed.

File: build.sbt [hunk 6]
```diff
@@ -4,7 +4,7 @@ crossScalaVersions := Seq("2.10.6", "2.11.7", "2.12.0-M3")
 val commonSettings: Seq[Setting[_]] = Seq(
   organization := "be.doeraene",
 
-  version := "0.1.1-SNAPSHOT",
+  version := "0.2.0-SNAPSHOT",
```
Comment (sjrd) [inline build.sbt]: I don't see any reason for bumping to 0.2.0, as everything stays backwards source and binary compatible. This change should be reverted, as well as those in `sbt-plugin-test/build.sbt` and `sbt-plugin-test/project/plugins.sbt`.

File: sbt-scalajs-reflection/src/main/scala/be/doeraene/sjsreflect/sbtplugin/ReflectSelectors.scala [hunk 7]
```diff
@@ -27,6 +27,11 @@ object ReflectSelectors {
   /** Allows to find the selected by its fully qualified name. */
   private[sbtplugin] case object ReflectClassByName extends Operation
 
+  /**
+    * Allows classes selected by this modifier to be enumerated as an array
+    */
+  private[sbtplugin] case object ReflectEnumerateClass extends Operation
```
Comment (sjrd) [inline sbt-scalajs-reflection/src/main/scala/be/doeraene/sjsreflect/sbtplugin/ReflectSelectors.scala:31]: Please follow the style of comments. Either right above if single-line, or right below if multi-line.

File: sbt-scalajs-reflection/src/main/scala/be/doeraene/sjsreflect/sbtplugin/ReflectionLinker.scala [hunk 8]
```diff
@@ -162,6 +162,14 @@ final class ReflectionLinker(underlying: GenLinker,
       JSObjectConstr(items)
     }
 
+    def makeClassArray()(implicit pos: ir.Position): Tree = {
+      val shouldEnable = shouldEnableOperation(ReflectEnumerateClass)
+
+      val items = infos.withFilter(shouldEnable).map(info => ClassOf(ClassType(info.encodedName)))
```
Comment (sjrd) [inline sbt-scalajs-reflection/src/main/scala/be/doeraene/sjsreflect/sbtplugin/ReflectionLinker.scala]: Consider following the style used in `makeClassesByName()`, i.e., using a for comprehension:

``` scala
       val items = for {
         info <- infos
         if shouldEnable(info)
       } yield {
         ClassOf(ClassType(info.encodedName))
       }
```

File: src/main/scala/be/doeraene/sjsreflect/Reflect.scala [hunk 9]
```diff
@@ -45,6 +45,13 @@ object Reflect {
   def getClassForName(fqcn: String): Option[Class[_]] =
     classesByName.get(fqcn)
 
+  /**
+    * returns an array of all classes specified by the `reflectEnumerateClass()` operator
+    * (and descendants if the  `selectDescendentClasses` selector is used)
+    */
+  def enumerateClasses: js.Array[Class[_]] = classArray
```
Comment (sjrd) [inline src/main/scala/be/doeraene/sjsreflect/Reflect.scala]: Same comment as above for the Scaladoc style.

Also, returning the same array is not safe, as arrays are mutable. A client could modify the returned array and pollute everyone else's versions. You should make a copy. Also consider returning a (Scala) `Array[Class[_]]` rather than a `js.Array[_]`, to be more consistent with the rest of the API (notably `getDeclaredConstructors`). So basically:

``` scala
def enumerateClasses: Array[Class[_]] = classArray.toArray
```

(since `toArray` makes a copy already)

File: src/main/scala/be/doeraene/sjsreflect/Reflect.scala [hunk 10]
```diff
@@ -45,6 +45,13 @@ object Reflect {
...
+
+
```
Comment (sjrd) [inline src/main/scala/be/doeraene/sjsreflect/Reflect.scala:50]: Double blank line. (there should be only one)

**Review by sjrd — CHANGES_REQUESTED** [review]
Sorry for the delay in reviewing this.

I think the API should be good now. But we need tests. Look at other tests in https://github.com/sjrd/scalajs-reflect/blob/master/src/test/scala/be/doeraene/sjsreflect/ReflectTest.scala for inspiration. Be as thorough as possible.

File: src/main/scala/be/doeraene/sjsreflect/Reflect.scala [hunk 11]
```diff
@@ -23,6 +21,8 @@ object Reflect {
 
   private[Reflect] def makeClassesByName(): js.Dictionary[Class[_]] = stub()
 
+  private[Reflect] def makeClassArray(): Array[Class[_]] = stub()
```
Comment (sjrd) [inline src/main/scala/be/doeraene/sjsreflect/Reflect.scala]: This one should still be a `js.Array`, otherwise the `JSArrayConstr` used in the linker plugin does not comply to. That's why the build failed.

(The other one, the public method, should return `Array[Class[_]]` as discussed before.)

File: src/test/scala/be/doeraene/sjsreflect/ReflectTest.scala [hunk 12]
```diff
@@ -65,6 +71,29 @@ class ReflectTest {
   }
 
   @Test
+  def enumerateClasses(): Unit = {
+    assertTrue(
+      Reflect.enumerateClasses.contains(classOf[EnumeratedClass])
+    )
```
Comment (sjrd) [inline src/test/scala/be/doeraene/sjsreflect/ReflectTest.scala]: To follow the coding style, either write:

``` scala
assertTrue(Reflect.enumerateClasses.contains(classOf[EnumeratedClass]))
```

if it fits within 80 characters, otherwise the following:

``` scala
assertTrue(
    Reflect.enumerateClasses.contains(classOf[EnumeratedClass]))
```

File: README.md [hunk 13]
```diff
@@ -49,14 +49,17 @@ at the end of its full name, e.g., `"foo.Bar$"`.
 Available operations are:
 
 * `reflectClassByName()`: enables `Reflect.getClassForName`
-* `reflectDeclaredConstructors()`: enables `Reflect.getDeclaredConstructors`
+* `reflectEnumerateClass()` enables `Reflect.enumerateClasses`
+* `reflectDeclaredConstructors()`: enables `Reflect.getDeclaredConstructors`
```
Comment (sjrd) [inline README.md]: There's a trailing whitespace on this line.

File: README.md [hunk 14]
```diff
@@ -49,14 +49,17 @@ at the end of its full name, e.g., `"foo.Bar$"`.
...
+* `reflectDeclaredConstructors()`: enables `Reflect.getDeclaredConstructors` 
 * `reflectModuleAccessor()`: enables `Reflect.loadModule`
 
 ## API
 
 * `be.doeraene.sjsreflect.Reflect`
   * `Reflect.getClassForName(fqcn: String): Class[_]`:
     find a class by its name
+  * `Reflect.enumerateClasses: Array[Class[_]]`:
+    an array of the specified classes
```
Comment (sjrd) [inline README.md]: There's a trailing whitespace on this line.

**Review by sjrd — CHANGES_REQUESTED** [review]
That's all I have.

When you've addressed these last small comments, please squash all your commits in a single commit with the following commit message:

```
Add Reflect.enumerateClasses to enumerate all classes selected with reflectEnumerateClass().
```

**Comment by AugustNagro** [comment]
Made the changes requested. 

I apologize for the sloppy pull request

**Review by sjrd — APPROVED** [review]
Thanks.
